simulator/ssh-transfer.go

```diff
@@ -0,0 +1,158 @@
+		} else {
+			payloadSize = toSend
+		}
+		bytes = make([]byte, payloadSize)
+		rand.Read(bytes)
+		// Don't care about the actual response here.  Just the error code.
+		_, err := c.SendWrite(handle, i*buffSize, &bytes)
+		if err != nil {
+			return fmt.Errorf("failed transfer: %v", err)
+		}
+		toSend -= payloadSize
+	}
+	return nil
+}
+
+// Simulate an ssh/sftp file transfer.
+func (s *SSHTransfer) Simulate(ctx context.Context, dst string) error {
+	// Auth.
+	signer, err := simssh.NewSignerFromKey()
+	if err != nil {
+		return err
+	}
+	// Create a Client that's ready to use for SSH/SFTP transfers.
+	c, err := simssh.NewClient("alphasoc", s.src, dst, signer)
+	if err != nil {
+		return err
+	}
+	defer c.Teardown()
+	// Init/Version.
+	initResp, err := c.SendInit()
+	if err != nil {
+		return err
+	}
+	// TODO: Do we really care about version mismatches?  From the sftp spec, a 3 can be
+	// followed by some form of version negotiaion.
+	if initResp.Version != simssh.ClientVer {
+		return fmt.Errorf("server version mismatch, expecting %v, received %v",
+			simssh.ClientVer, initResp.Version)
+	}
+	// Open a dummy file for writing and grab the handle returned by the server.  If used
+	// with the sandbox SFTP server, no filesystem writes will actually occurr.
+	openResp, err := c.SendOpen("thereisnofile", os.O_CREATE)
+	if err != nil {
+		return err
+	}
+	handle := openResp.Handle
+	// Write s.sendSize bytes, checking for any write errors.
+	err = writeRandom(c, handle, s.sendSize)
+	if err != nil {
+		return err
+	}
+	// Close the handle.  We don't care about the response, just the error.
+	_, err = c.SendClose(handle)
+	if err != nil {
+		return err
+	}
+	// Success.
+	return nil
+}
+
+// Cleanup is a no-op.
+func (s *SSHTransfer) Cleanup() {}
+
+// parseScope parses the commandline portion (if supplied) after the module name.
+//   ie. flightsim run ssh-transfer:this-part-is-scope:and-can-contain-more
+// For the moment, only send size is parsed, but ultimately we also want to pass
+// destination host and port.  Returns a string representation of the destination
+// host (currently ""), the send size as a ByteSize, and an error.
+func parseScope(scope string) ([]string, bytesize.ByteSize, error) {
+
+	// scope can be "", in which case, apply defaults.
+	if scope == "" {
+		return defaultTargetHosts, defaultSendSize, nil
+	}
+	// scope may contain just the send size (ie. a lack of futher ":" separators
+	// present in the string).
+	var sendSize bytesize.ByteSize
+	var err error
+	if !strings.Contains(scope, ":") {
+		sendSize, err = bytesize.Parse(scope)
+		if err != nil {
+			return []string{""},
+				bytesize.ByteSize(0),
+				fmt.Errorf("invalid command line: %v", err)
+		}
+		return defaultTargetHosts, sendSize, nil
+	}
+	// TODO scope may contain more information, separated by ":", perhaps as key-value
+	// pairs.  For now, not supported.
+	return []string{""}, bytesize.ByteSize(0), fmt.Errorf("invalid command line: %v", scope)
```

Comment:
I don't think "invalid command line" is a valid string here. This function can be called by anyone anywhere and doesn't need to parse command lines, so it shouldn't be so specific.

Reply:
Same applies to line 140. And really we should have two exit points sharing the same string here. Interestingly you pass `err` in first case and `scope` in another. This should be unified, pass both `scope` and `err` and used only once in the code without repetition.

Reply:
Also, just return `nil` instead of `[]string{""}` in case of error.